pkg/daemon/file_writers.go

```diff
@@ -0,0 +1,302 @@
+	t, err := renameio.TempFile(dir, fpath)
+	if err != nil {
+		return err
+	}
+	defer t.Cleanup()
+	// Set permissions before writing data, in case the data is sensitive.
+	if err := t.Chmod(fileMode); err != nil {
+		return err
+	}
+	w := bufio.NewWriter(t)
+	if _, err := w.Write(b); err != nil {
+		return err
+	}
+	if err := w.Flush(); err != nil {
+		return err
+	}
+	if uid != -1 && gid != -1 {
+		if err := t.Chown(uid, gid); err != nil {
+			return err
+		}
+	}
+	return t.CloseAtomicallyReplace()
+}
+
+// write dropins to disk
+func writeDropins(u ign3types.Unit, systemdRoot string, isCoreOSVariant bool) error {
+	for i := range u.Dropins {
+		dpath := filepath.Join(systemdRoot, u.Name+".d", u.Dropins[i].Name)
+		if u.Dropins[i].Contents == nil || *u.Dropins[i].Contents == "" {
+			glog.Infof("Dropin for %s has no content, skipping write", u.Dropins[i].Name)
+			if _, err := os.Stat(dpath); err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return err
+			}
+			glog.Infof("Removing %q, updated file has zero length", dpath)
+			if err := os.Remove(dpath); err != nil {
+				return err
+			}
+			continue
+		}
+
+		glog.Infof("Writing systemd unit dropin %q", u.Dropins[i].Name)
+		if _, err := os.Stat(withUsrPath(dpath)); err == nil &&
+			isCoreOSVariant {
+			if err := createOrigFile(withUsrPath(dpath), dpath); err != nil {
+				return err
+			}
+		}
+		if err := writeFileAtomicallyWithDefaults(dpath, []byte(*u.Dropins[i].Contents)); err != nil {
+			return fmt.Errorf("failed to write systemd unit dropin %q: %w", u.Dropins[i].Name, err)
+		}
+
+		glog.V(2).Infof("Wrote systemd unit dropin at %s", dpath)
+	}
+
+	return nil
+}
+
+// writeFiles writes the given files to disk.
+// it doesn't fetch remote files and expects a flattened config file.
+func writeFiles(files []ign3types.File) error {
+	for _, file := range files {
+		glog.Infof("Writing file %q", file.Path)
+
+		// We don't support appends in the file section, so instead of waiting to fail validation,
+		// let's explicitly fail here.
+		if len(file.Append) > 0 {
+			return fmt.Errorf("found an append section when writing files. Append is not supported")
+		}
+
+		decodedContents, err := ctrlcommon.DecodeIgnitionFileContents(file.Contents.Source, file.Contents.Compression)
+		if err != nil {
+			return fmt.Errorf("could not decode file %q: %w", file.Path, err)
+		}
+
+		mode := defaultFilePermissions
+		if file.Mode != nil {
+			mode = os.FileMode(*file.Mode)
+		}
+
+		// set chown if file information is provided
+		uid, gid, err := getFileOwnership(file)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve file ownership for file %q: %w", file.Path, err)
+		}
+		if err := createOrigFile(file.Path, file.Path); err != nil {
+			return err
+		}
+		if err := writeFileAtomically(file.Path, decodedContents, defaultDirectoryPermissions, mode, uid, gid); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeUnit writes s systemd unit and its dropins to disk
```

Comment:
nit: did you mean to write `writes a* systemd unit`?

Reply:
Done!